File: src/maincontroller.cpp
#include "maincontroller.h"

MainController::MainController(QObject *parent)
    : QObject{parent},
      m_mainContentQml("Home.qml"),
      m_activeId("HOME"),
      m_counter(0)
{
    m_timer.setInterval(10);
    m_timer.setSingleShot(false);

    // Using these two lines, it runs indefinitely
//    auto h = new QmlFileWrapper("Home.qml", this);
//    auto o = new QmlFileWrapper("Other.qml", this);

    // Using these two lines, it ends after 155 iterations
    auto h = new QmlFileWrapper("Home.qml");
    auto o = new QmlFileWrapper("Other.qml");

    QObject::connect(h, &QObject::destroyed, this, &MainController::onDestroy);
    QObject::connect(o, &QObject::destroyed, this, &MainController::onDestroy);
    m_qmlWrappers.insert(QString("HOME"), h);
    m_qmlWrappers.insert(QString("OTHER"), o);

    m_activeWrapper = get_wrapper_object_by_name(m_activeId);
}

void MainController::onDestroy(QObject* obj)
{
    qDebug("Trashing object at %p", obj);
}

const QString &MainController::mainContentQml() const
{
    return m_activeWrapper->qmlPath();
}

void MainController::startup()
{
    QObject::connect(&m_timer, &QTimer::timeout, this, &MainController::toggleScreen);
    m_timer.start();
}

void MainController::shutdown()
{
    QObject::disconnect(&m_timer, &QTimer::timeout, this, &MainController::toggleScreen);
    m_timer.stop();
    qDebug("Stopping after %d iterations.", m_counter);
}

QmlFileWrapper* MainController::get_wrapper_object_by_name(QString next_id)
{
    auto i = m_qmlWrappers.find(next_id);
    return i.value();
}

void MainController::goToQmlByName(QString next_id)
{
    m_activeWrapper = get_wrapper_object_by_name(next_id);
    emit mainContentQmlChanged();
}

void MainController::toggleScreen()
{
    m_counter++;
    if (m_activeId == QString("HOME")) {
        m_activeId = QString("OTHER");
    }
    else {
        m_activeId = QString("HOME");
    }
    goToQmlByName(m_activeId);
}

File: errortest/main_controller.py
from PySide6.QtCore import QObject, Property, Slot, Signal, QTimer
from typing import Optional

from .qml_file_wrapper import QmlFileWrapper


class MainController(QObject):
    main_content_qml_changed = Signal()

    def __init__(self, parent=None):
        super().__init__(parent)
        self._app = parent
        self._qml_wrappers = {
            "HOME": QmlFileWrapper('Home.qml'),
            "OTHER": QmlFileWrapper('Other.qml')
        }
        self._active_id = "HOME"
        self._active_wrapper: QmlFileWrapper = self._qml_wrappers[self._active_id]
        self._counter = 0
        self._timer = QTimer()
        self._timer.setInterval(10)
        self._timer.setSingleShot(False)
        self._timer.timeout.connect(self._toggle_screen)

    @Property(str, notify=main_content_qml_changed)
    def main_content_qml(self) -> str:
        return self._active_wrapper.qml_path

    def startup(self):
        self._timer.start()

    def shutdown(self):
        print(f"Stopping after {self._counter} iterations.")

    @Slot(str, result=QmlFileWrapper)
    def get_wrapper_object_by_name(self, screen_name: str) -> Optional[QmlFileWrapper]:
        return self._qml_wrappers[screen_name.upper()]

    @Slot(str)  # QML will only send a string
    def go_to_qml_by_name(self, next_id: str) -> None:
        self._active_wrapper = self.get_wrapper_object_by_name(next_id)
        self.main_content_qml_changed.emit()

    def _toggle_screen(self):
        self._counter = self._counter + 1
        if self._active_id == "HOME":
            self._active_id = "OTHER"
        else:
            self._active_id = "HOME"

        self.go_to_qml_by_name(self._active_id)
